content/en/docs/concepts/storage/storage-limits.md

```diff
@@ -78,4 +53,35 @@ Refer to the [CSI specifications](https://github.com/container-storage-interface
 
 * For volumes managed by in-tree plugins that have been migrated to a CSI driver, the maximum number of volumes will be the one reported by the CSI driver.
 
+## Custom limits
+
+You can change these limits by setting the value of the
+`KUBE_MAX_PD_VOLS` environment variable, and then starting the scheduler.
+CSI drivers might have a different procedure, see their documentation
+on how to customize their limits.
+
+Use caution if you set a limit that is higher than the default limit. Consult
+the cloud provider's documentation to make sure that Nodes can actually support
+the limit you set.
+
+The limit applies to the entire cluster, so it affects all Nodes.
+
+
+## Kubernetes default limits
+
+{{< note >}}
+Dynamic volume limits is the recommended approach that all storage plugins should use
+{{< /note >}}
+
+The Kubernetes scheduler has default limits on the number of volumes
```

Comment:
Note that these are all deprecated and will be replaced by dynamic limits

Reply:
:+1:

Reply:
Hi @msau42 Should I edit the Note section and add the deprecation message or remove the "Default Limits" section entirely?

Reply:
@bertinatto I believe we don't fallback to these limits anymore and we only use the dynamic limits? Can you confirm?

Reply:
The scheduler falls-back to these limits if it doesn't find a `CSINode` object for the node, which happens if there's no CSI driver installed for that particular cloud (or migration is off).

Reply:
My understanding is these scenarios will use dynamic limits:
1. Pod requests in-tree volume and CSI migration is off
1. Pod requests in-tree volume + CSI migration is on + driver is installed
1. Pod requests CSI volume + driver is installed


Could you clarify the behavior in these scenarios?
1. Pod requests intree + CSI migration on + driver not installed
1. Pod requests CSI + driver not installed

Reply:
> My understanding is these scenarios will use dynamic limits:
> 
>     1. Pod requests in-tree volume and CSI migration is off

I don't think so, the CSI predicate (the one that checks for dynamic volume limits) won't count the volume if it hasn't been migrated.

https://github.com/kubernetes/kubernetes/blob/1cbda2493d2744af0d0e52b008976af0fc1d3a5e/pkg/scheduler/framework/plugins/nodevolumelimits/csi.go#L194


>     2. Pod requests in-tree volume + CSI migration is on + driver is installed

Yes

>     3. Pod requests CSI volume + driver is installed

Yes
> 
> Could you clarify the behavior in these scenarios?
> 
>     1. Pod requests intree + CSI migration on + driver not installed

It should **not** use dynamic volume limits, **unless** a CSINode exists and has the limits value set.

https://github.com/kubernetes/kubernetes/blob/1cbda2493d2744af0d0e52b008976af0fc1d3a5e/pkg/scheduler/framework/plugins/nodevolumelimits/utils.go#L34

>     2. Pod requests CSI + driver not installed

We don't check if the driver is installed (it was decided to drop this requirement, can't remember the details).